s3transfer/download.py

```diff
@@ -475,3 +558,38 @@ def __init__(self, transfer_coordinator, main_kwargs=None,
 
     def _main(self):
         pass
+
+
+class DeferQueue(object):
+    """IO queue that defers write requests until they are queued sequentially.
+
+    This class is used to track IO data for a *single* fileobj.
+
+    You can send data to this queue, and it will defer any IO write requests
+    until it has the next contiguous block available (starting at 0).
+
+    """
+    def __init__(self):
+        self._writes = []
+        self._next_offset = 0
+
+    def request_writes(self, offset, data):
+        """Request any available writes given new incoming data.
+
+        You call this method by providing new data along with the
+        offset associated with the data.  If that new data unlocks
+        any contiguous writes that can now be submitted, this
+        method will return all applicable writes.
+
+        This is done with 1 method call so you don't have to
+        make two method calls (put(), get()) which acquires a lock
+        each method call.
+
+        """
+        writes = []
+        heapq.heappush(self._writes, (offset, data))
+        while self._writes and self._writes[0][0] == self._next_offset:
```

Comment:
Can you explain how retries are handled? For example, we start adding data to the heap and that bumps up the ``self._next_offset``. Then let's say that the download got retried and we start at the beginning again with an offset less than the ``self._next_offset``. It looks like the queue will never be cleared because the reset offset will be smaller than the expected offse and the while loop will never be entered. To fix this, I think you need to check if the given offset is less than the _next_offset and discard it if so.